### PR TITLE
Work on Networks part 1

### PR DESCRIPTION
**Pull request by bokobza**
This is the beginning of the work aiming to streamline how Networks are created.

* Removed NetworkBuilder in favor of using good ol' property assignements, while still keeping immutability of the Network object
* Remove interdependance of the Network. Before, every initialization of a Network was calling another Network in order to copy some of its data, resulting in a cascade of unnecessary intializations. While it's not such a big deal, these dependencies have been removed.
* Added some tests that checks that the Networks have been properly initialized.
* Tidied up the Init Network methods so they all look more or less the same.

There is more to do but I need to work on Integration tests first, I've wasted more time testing this than actually writing any code.

Unit tests:
image

Bitcoin Main from scratch:
image

Bitcoin test from scratch:
image

Stratis Main from scratch:

image

Stratis Test from scratch:

image

### CODE REVIEW

File: src/NBitcoin/Network.cs [hunk 1]
```diff
@@ -235,19 +235,20 @@ public virtual Consensus Clone()
...
-        private readonly Dictionary<int, CheckpointInfo> checkpoints = new Dictionary<int, CheckpointInfo>();
         private Block genesis;
-        private Consensus consensus = new Consensus();
-
+        
         private Network()
         {
-            this.seeds = new List<DNSSeedData>();
-            this.fixedSeeds = new List<NetworkAddress>();
+            this.DNSSeeds = new List<DNSSeedData>();
```
Comment (noescape00) [inline src/NBitcoin/Network.cs]: Can you please add comments for those properties? (I know it's kind of not the aim of this PR but if not in this PR- no one will do it for months)

Comment (bokobza) [reply src/NBitcoin/Network.cs]: yeah you're right. 
I was thinking this class' properties are some of the most important ones we have in the code. Every user using our node will have to understand what they mean and what they're for. That's also why I didn't pay attention to them at this stage.
I'll do that in part 2 if that's ok, I really want to get going on the Integration Tests.

File: src/NBitcoin/Network.cs [hunk 2]
```diff
@@ -315,77 +322,29 @@ public byte[] MagicBytes
...
         }
-
-        public uint Magic => this.magic;
-
-        private static readonly ConcurrentDictionary<string, Network> NetworksContainer = new ConcurrentDictionary<string, Network>();
-
-        internal static Network Register(NetworkBuilder builder)
+        
+        /// <summary>
+        /// Register an immutable Network instance so it is queriable through Network.GetNetwork(string name) and Network.GetNetworks().
```
Comment (noescape00) [inline src/NBitcoin/Network.cs]: typo, htmldoc, redundant return

fixed:

```
/// <summary>
/// Register an immutable <see cref="Network"/> instance so it is queryable through <see cref="GetNetwork(string)"/> and <see cref="GetNetworks()"/>.
/// </summary>
```

**Comment by bokobza** [comment]
Thanks @noescape00 !

**Comment by noescape00** [comment]
@dangershony I'm happy for this PR to go in. Won't push since you're marked as a reviewer, so please push when you check it.

File: src/NBitcoin/Network.cs [hunk 3]
```diff
@@ -278,7 +279,7 @@ public PubKey AlertPubKey
 
         public int DefaultPort { get; private set; }
 
-        public Consensus Consensus => this.consensus;
+        public Consensus Consensus { get; private set; }
```
[the remark this answers is not in the document]
Comment (bokobza) [reply src/NBitcoin/Network.cs]: mmh don't think so. this was a straight change.

```
// both at class level
private Consensus consensus = new Consensus();
public Consensus Consensus => this.consensus;
```

File: src/NBitcoin/Networks.cs [hunk 4]
```diff
@@ -315,43 +315,45 @@ private static Network InitReg()
...
-            consensus.BuriedDeployments[BuriedDeployments.BIP66] = 363725;
-            consensus.BIP34Hash = new uint256("0x000000000000024b89b42a942fe0d9fea3bb44ab7bd1b19115dd6a759c0808b8");
-            consensus.PowLimit = new Target(new uint256("00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"));
-            consensus.PowTargetTimespan = TimeSpan.FromSeconds(14 * 24 * 60 * 60); // two weeks
-            consensus.PowTargetSpacing = TimeSpan.FromSeconds(10 * 60);
-            consensus.PowAllowMinDifficultyBlocks = false;
-            consensus.PowNoRetargeting = false;
-            consensus.RuleChangeActivationThreshold = 1916; // 95% of 2016
-            consensus.MinerConfirmationWindow = 2016; // nPowTargetTimespan / nPowTargetSpacing
+            network.Consensus.NetworkOptions = new NetworkOptions();
```
Comment (quantumagi) [inline src/NBitcoin/Networks.cs]: All "network.Consensus... = ..." could be moved to a "StratisConsensus" class, derived from Consensus.

**Comment by bokobza** [comment]
So actually, I had to remove BIP9Deployments and BuriedDeployments from the stratis networks as they are not used (I checked in StratisX).
The previous code added them to Consensus but didn't copy them when Consensus was cloned into the Network object.
This has now been rectified.

All tests are passing. I also ran my NetworkTests on the old code and it's all in top shape, meaning the Network objects that are created with the new code are exactly similar with the old code 👍 

Also running the 4 nodes here, my computer is purring nicely.
@dangershony @Aprogiena @noescape00 @quantumagi

File: src/NBitcoin/Networks.cs (hunk 4 above)
Comment (quantumagi) [reply src/NBitcoin/Networks.cs]: Yep but what I meant is that that line plus all the following ones of the form `network.Consensus... = ...` could be moved to a separate class - e.g. StratisConsensus. Not something we need to do right now though.

**Pull request merged**